nexus2cbf test optimization-dependent crashes

I am building and CBFLib 0.9.6 from the distribution package available at SourceForge, on CentOS Linux 8 (x86_64), using the distribution's GCC toochain version 8.3.1.  The library, wrappers, and example programs all built successfully, but when I ran the tests, nexus2cbf crashed:
```
LD_LIBRARY_PATH=/home/jbolling/tmp/cbflib-CBFlib-0.9.6/solib:$LD_LIBRARY_PATH;export LD_LIBRARY_PATH; cd /home/jbolling/tmp/cbflib-CBFlib-0.9.6/minicbf_test; time /home/jbolling/tmp/cbflib-CBFlib-0.9.6/bin/nexus2cbf \
-o i19-1.cbf i19-1.h5
HDF5-DIAG: Error detected in HDF5 (1.8.18) thread 0:
  #000: H5D.c line 342 in H5Dopen2(): no name
    major: Invalid arguments to routine
    minor: Bad value
HDF5-DIAG: Error detected in HDF5 (1.8.18) thread 0:
  #000: H5L.c line 1183 in H5Literate(): link iteration failed
    major: Symbol table
    minor: Iteration failed
  #001: H5Gint.c line 844 in H5G_iterate(): error iterating over links
    major: Symbol table
    minor: Iteration failed
  #002: H5Gobj.c line 708 in H5G__obj_iterate(): can't iterate over symbol table
    major: Symbol table
    minor: Iteration failed
  #003: H5Gstab.c line 566 in H5G__stab_iterate(): iteration operator failed
    major: Symbol table
    minor: Can't move to next iterator location
  #004: H5B.c line 1221 in H5B_iterate(): B-tree iteration failed
    major: B-Tree node
    minor: Iteration failed
  #005: H5B.c line 1177 in H5B_iterate_helper(): B-tree iteration failed
    major: B-Tree node
    minor: Iteration failed
  #006: H5Gnode.c line 1039 in H5G__node_iterate(): iteration operator failed
    major: Symbol table
    minor: Can't move to next iterator location
HDF5-DIAG: Error detected in HDF5 (1.8.18) thread 0:
  #000: H5L.c line 1183 in H5Literate(): link iteration failed
    major: Symbol table
    minor: Iteration failed
  #001: H5Gint.c line 844 in H5G_iterate(): error iterating over links
    major: Symbol table
    minor: Iteration failed
  #002: H5Gobj.c line 708 in H5G__obj_iterate(): can't iterate over symbol table
    major: Symbol table
    minor: Iteration failed
  #003: H5Gstab.c line 566 in H5G__stab_iterate(): iteration operator failed
    major: Symbol table
    minor: Can't move to next iterator location
  #004: H5B.c line 1221 in H5B_iterate(): B-tree iteration failed
    major: B-Tree node
    minor: Iteration failed
  #005: H5B.c line 1177 in H5B_iterate_helper(): B-tree iteration failed
    major: B-Tree node
    minor: Iteration failed
  #006: H5Gnode.c line 1039 in H5G__node_iterate(): iteration operator failed
    major: Symbol table
    minor: Can't move to next iterator location
HDF5-DIAG: Error detected in HDF5 (1.8.18) thread 0:
  #000: H5L.c line 1183 in H5Literate(): link iteration failed
    major: Symbol table
    minor: Iteration failed
  #001: H5Gint.c line 844 in H5G_iterate(): error iterating over links
    major: Symbol table
    minor: Iteration failed
  #002: H5Gobj.c line 708 in H5G__obj_iterate(): can't iterate over symbol table
    major: Symbol table
    minor: Iteration failed
  #003: H5Gstab.c line 566 in H5G__stab_iterate(): iteration operator failed
    major: Symbol table
    minor: Can't move to next iterator location
  #004: H5B.c line 1221 in H5B_iterate(): B-tree iteration failed
    major: B-Tree node
    minor: Iteration failed
  #005: H5B.c line 1177 in H5B_iterate_helper(): B-tree iteration failed
    major: B-Tree node
    minor: Iteration failed
  #006: H5Gnode.c line 1039 in H5G__node_iterate(): iteration operator failed
    major: Symbol table
    minor: Can't move to next iterator location
An error occured, will not try to write the file 'i19-1.cbf'
Time to convert 'i19-1.cbf': 0.006s

real    0m0.008s
user    0m0.006s
sys     0m0.002s
make: *** [Makefile_LINUX_64:2230: extra] Error 1
```

I found, however, that if I used the default Makefile instead of Makefile_LINUX_64, then the same test ran successfully.  Ultimately, I traced the issue to optimization level.  nexus2cbf works if everything is built with optimization `-O3` (default Makefile), but not if everything is built with optimization `-O2` (Makefile_LINUX_64 and Makefile_LINUX).

There is a clear workaround here, of course, but optimization-dependent misbehavior is usually a sign of a deeper problem, especially when it is the more aggressive optimization level that is required for (apparently) correct behavior.

Dear John,
  The hdf5 library is, unfortunately, fragile.  Some combinations work on
some systems and not on others.  Things got worse
with hdf5-1.10.1 through hdf5-1.10.4, then a little better with hdf5-1.10.5
and still better with hdf5-1.10.6.  Now hdf5-1.12 is
out, so I'll be investigating that next.  Other destabilizing issues are
use or no use of dlfnc, gcc and gfortran versions or
use of intel compilers.
  Yes there is a deep problem here.  Got a student who likes a challenge?
  Regards,
    Herbert

On Wed, Apr 29, 2020 at 6:29 PM John Bollinger <notifications@github.com>
wrote:

> I am building and CBFLib 0.9.6 from the distribution package available at
> SourceForge, on CentOS Linux 8 (x86_64), using the distribution's GCC
> toochain version 8.3.1. The library, wrappers, and example programs all
> built successfully, but when I ran the tests, nexus2cbf crashed:
>
> LD_LIBRARY_PATH=/home/jbolling/tmp/cbflib-CBFlib-0.9.6/solib:$LD_LIBRARY_PATH;export LD_LIBRARY_PATH; cd /home/jbolling/tmp/cbflib-CBFlib-0.9.6/minicbf_test; time /home/jbolling/tmp/cbflib-CBFlib-0.9.6/bin/nexus2cbf \
> -o i19-1.cbf i19-1.h5
> HDF5-DIAG: Error detected in HDF5 (1.8.18) thread 0:
>   #000: H5D.c line 342 in H5Dopen2(): no name
>     major: Invalid arguments to routine
>     minor: Bad value
> HDF5-DIAG: Error detected in HDF5 (1.8.18) thread 0:
>   #000: H5L.c line 1183 in H5Literate(): link iteration failed
>     major: Symbol table
>     minor: Iteration failed
>   #001: H5Gint.c line 844 in H5G_iterate(): error iterating over links
>     major: Symbol table
>     minor: Iteration failed
>   #002: H5Gobj.c line 708 in H5G__obj_iterate(): can't iterate over symbol table
>     major: Symbol table
>     minor: Iteration failed
>   #003: H5Gstab.c line 566 in H5G__stab_iterate(): iteration operator failed
>     major: Symbol table
>     minor: Can't move to next iterator location
>   #004: H5B.c line 1221 in H5B_iterate(): B-tree iteration failed
>     major: B-Tree node
>     minor: Iteration failed
>   #005: H5B.c line 1177 in H5B_iterate_helper(): B-tree iteration failed
>     major: B-Tree node
>     minor: Iteration failed
>   #006: H5Gnode.c line 1039 in H5G__node_iterate(): iteration operator failed
>     major: Symbol table
>     minor: Can't move to next iterator location
> HDF5-DIAG: Error detected in HDF5 (1.8.18) thread 0:
>   #000: H5L.c line 1183 in H5Literate(): link iteration failed
>     major: Symbol table
>     minor: Iteration failed
>   #001: H5Gint.c line 844 in H5G_iterate(): error iterating over links
>     major: Symbol table
>     minor: Iteration failed
>   #002: H5Gobj.c line 708 in H5G__obj_iterate(): can't iterate over symbol table
>     major: Symbol table
>     minor: Iteration failed
>   #003: H5Gstab.c line 566 in H5G__stab_iterate(): iteration operator failed
>     major: Symbol table
>     minor: Can't move to next iterator location
>   #004: H5B.c line 1221 in H5B_iterate(): B-tree iteration failed
>     major: B-Tree node
>     minor: Iteration failed
>   #005: H5B.c line 1177 in H5B_iterate_helper(): B-tree iteration failed
>     major: B-Tree node
>     minor: Iteration failed
>   #006: H5Gnode.c line 1039 in H5G__node_iterate(): iteration operator failed
>     major: Symbol table
>     minor: Can't move to next iterator location
> HDF5-DIAG: Error detected in HDF5 (1.8.18) thread 0:
>   #000: H5L.c line 1183 in H5Literate(): link iteration failed
>     major: Symbol table
>     minor: Iteration failed
>   #001: H5Gint.c line 844 in H5G_iterate(): error iterating over links
>     major: Symbol table
>     minor: Iteration failed
>   #002: H5Gobj.c line 708 in H5G__obj_iterate(): can't iterate over symbol table
>     major: Symbol table
>     minor: Iteration failed
>   #003: H5Gstab.c line 566 in H5G__stab_iterate(): iteration operator failed
>     major: Symbol table
>     minor: Can't move to next iterator location
>   #004: H5B.c line 1221 in H5B_iterate(): B-tree iteration failed
>     major: B-Tree node
>     minor: Iteration failed
>   #005: H5B.c line 1177 in H5B_iterate_helper(): B-tree iteration failed
>     major: B-Tree node
>     minor: Iteration failed
>   #006: H5Gnode.c line 1039 in H5G__node_iterate(): iteration operator failed
>     major: Symbol table
>     minor: Can't move to next iterator location
> An error occured, will not try to write the file 'i19-1.cbf'
> Time to convert 'i19-1.cbf': 0.006s
>
> real    0m0.008s
> user    0m0.006s
> sys     0m0.002s
> make: *** [Makefile_LINUX_64:2230: extra] Error 1
>
> I found, however, that if I used the default Makefile instead of
> Makefile_LINUX_64, then the same test ran successfully. Ultimately, I
> traced the issue to optimization level. nexus2cbf works if everything is
> built with optimization -O3 (default Makefile), but not if everything is
> built with optimization -O2 (Makefile_LINUX_64 and Makefile_LINUX).
>
> There is a clear workaround here, of course, but optimization-dependent
> misbehavior is usually a sign of a deeper problem, especially when it is
> the more aggressive optimization level that is required for (apparently)
> correct behavior.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/yayahjb/cbflib/issues/15>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABB6EAJ4WNYGBBP6EDSZ3YLRPCS47ANCNFSM4MUGFKYQ>
> .
>

--- a/src/cbf.c
+++ b/src/cbf.c
@@ -8395,9 +8395,9 @@ int cbf_mpint_rightshift_acc(unsigned int * acc, size_t acsize, int shift) {
   
   extrabits = 0;
 
-  if (acc[acsize-1]&sign) extrabits = (~0)<<(sizeof(unsigned int)*CHAR_BIT-shift);
+  if (acc[acsize-1]&sign) extrabits = (~0u)<<(sizeof(unsigned int)*CHAR_BIT-shift);
 
-  mask = ~((~0)<<(sizeof(unsigned int)*CHAR_BIT-shift));
+  mask = ~((~0u)<<(sizeof(unsigned int)*CHAR_BIT-shift));
 
   for (iint = acsize; iint; iint--) {
 

--- a/src/cbf_copy.c
+++ b/src/cbf_copy.c
@@ -2288,9 +2288,9 @@ extern "C" {
                 
                 index = elsize*(fastlow +indexmid*dimfast+indexslow*dimfast*dimmid);
                 
-                memmove(tdst,src+index,(1+fasthigh-fastlow)*elsize);
+                memmove(tdst,(char *)src+index,(1+fasthigh-fastlow)*elsize);
                 
-                tdst += (1+fasthigh-fastlow)*elsize;
+                tdst = (char *)tdst + (1+fasthigh-fastlow)*elsize;
                 
             }
         }

--- a/src/cbf_hdf5.c
+++ b/src/cbf_hdf5.c
@@ -304,6 +304,11 @@ extern "C" {
      int real,
      const char *byteorder);
 
+    int cbf_read_h5file_group(const cbf_handle handle,
+                        const cbf_h5handle h5handle,
+                        const unsigned long int flags,
+                        hid_t group);
+
     /* CBF column -- NeXus DataSet Mapping Tables
      applicable to categories that map to
      datasets or to attributes
@@ -17951,7 +17956,7 @@ _cbf_pilatusAxis2nexusAxisAttrs(h4data,units,depends_on,exsisItem,cmp)
 
         char svndate[] = CBF_SVN_DATE_STRING;
 
-        char buffer[140];
+        char buffer[140] = {0};
 
         int ii, irev, idate;
 
@@ -18033,7 +18038,7 @@ _cbf_pilatusAxis2nexusAxisAttrs(h4data,units,depends_on,exsisItem,cmp)
 
         char svndate[] = CBF_SVN_DATE_STRING;
 
-        char buffer[140];
+        char buffer[140] = {0};
 
         int ii, irev, idate;
 
@@ -19021,6 +19026,7 @@ _cbf_pilatusAxis2nexusAxisAttrs(h4data,units,depends_on,exsisItem,cmp)
             } else {
                 const htri_t vlstr = H5Tis_variable_str(data_type);
                 char * * lvalue = NULL;
+                char * llvalue = NULL;
                 int freelvalue = 0;
                 if (vlstr<0) {
                     cbf_debug_print("Couldn't check for a variable-length string");
@@ -19038,7 +19044,6 @@ _cbf_pilatusAxis2nexusAxisAttrs(h4data,units,depends_on,exsisItem,cmp)
                 } else {
                     /* I have a fixed-length string */
                     const size_t len = H5Tget_size(data_type);
-                    char * llvalue = NULL;
                     if (!len) {
                         cbf_debug_print("Couldn't get length of string");
                         error |= CBF_H5ERROR;

--- a/src/img.c
+++ b/src/img.c
@@ -1331,7 +1331,7 @@ int img_read_mar345header (img_handle img, FILE *file, int *org_data)
 
           *C = '\0';
           
-      for (C = C64+strlen(C64)-1; (C != C64-1 && *C == ' '); C-- ) *C = '\0';
+      for (C = C64+strlen(C64); (C > C64 && *--C == ' '); ) *C = '\0';
 
       C = C64 + strcspn (C64, " ");
 

